accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,584 @@
+- Existing usages of STJ such as by the SDK and Visual Studio that do not use `dynamic` also do not want to carry `System.Linq.Expressions.dll`.
+
+The preferred approach is to ensure ILLinker can remove the dependency to `System.Linq.Expressions.dll` when `dynamic` is not needed. This is achieved by having an opt-in method `EnableDynamicTypes()` on `JsonSerializerOptions` that roots all internal usages of `System.Linq.Expressions.dll`.
+
+## Proposed API to enable dynamic mode
+All classes to be located in STJ.dll.
+
+```cs
+namespace System.Text.Json
+{
+    public static class JsonSerializerOptions
+    {
+        // The root method that if not called, ILLinker will trim out the reference to `System.Linq.Expressions.dll`.
+        // Any System.Object-qualified type specified in a called to JsonSerializer.Deserialize() will create either a DynamicJsonArray, DynamicJsonObject or DynamicJsonValue.
+        public static void EnableDynamicTypes();
+    }
+}
+```
+
+Sample code to enable dynamic types:
+```cs
+    var options = new JsonSerializerOptions();
+    options.EnableDynamicTypes();
+
+    dynamic obj = JsonSerializer.Deserialize<JsonNode>("{\"MyProp\":42}", options);
+    int i = obj.MyProp;
+    Debug.Assert(i == 42);
+
+    string json = obj.Serialize();
+```
+
+### Varying the `T` in `Deserialize<T>`
+This
+```cs
+    dynamic obj = JsonSerializer.Deserialize<object>("{\"MyProp\":42}", options);
+```
+is equivalent to
+```cs
+    dynamic obj = JsonSerializer.Deserialize<dynamic>("{\"MyProp\":42}", options);
+```
+and these (assuming `options.EnableDynamicTypes()` was called)
+```cs
+    dynamic obj = JsonSerializer.Deserialize<DynamicJsonNode>("{\"MyProp\":42}", options);
+    dynamic obj = JsonSerializer.Deserialize<DynamicJsonObject>("{\"MyProp\":42}", options);
+```
+
+### Supporting non-dynamic to dynamic
+If a previously created non-dynamic `JsonNode` object needs dynamic support, an helper could be provided:
+```cs
+namespace System.Text.Json
+{
+    public static class JsonExtensions
+    {
+        public static dynamic ToDynamic(JsonNode this value);
+    }
+}
+```
+
+Called like this:
+```cs
+    JsonNode node = ... // a non-dynamic JsonNode
+    dynamic obj = node.ToDynamic();
+```
+
+The implementation of `ToDynamic()` is efficient since it will perform a shallow clone by copying the internal value (for `JsonValue`), array (for `JsonArray`) or dictionary (for `JsonObject`):
+```cs
+    JsonNode node = ... // a non-dynamic JsonNode
+    byte[] utf8 = node.SerializeToUtf8Bytes();
+    dynamic obj = JsonSerializer.Deserialize<JsonNode>(utf8, options);
+```
+
+# Writable DOM
+A deserialized DOM is based upon JSON token types (object, array, number, string, true, false) since the corresponding CLR types are not known, unlike deserializing against a strongly-typed POCO. This is consistent with `JsonElement` semantics as well.
```

Comment:
I haven't gotten a chance to read through all this yet, but can you help summarize how this design relates or differs from the existing effort (which had a hierarchy of types) from: https://github.com/dotnet/corefx/blob/master/src/System.Text.Json/docs/writable_json_dom_spec.md

Are there any fundamental differences, or is it mainly small updates over what we had initially?

Do we have learnings from the usability study incorporated into this design?

Reply:
These are the key differences that come to mind:
- Supports C# `dynamic`. The programming model for `dynamic` is consistent with non-dynamic cases. For example, to create a new JSON object, there is a `JsonObject` base class for non-dynamic cases and a derived `JsonDynamicObject` type for when dynamic is enabled.
- Has a sealed `JsonValue` class instead of separate "JsonNumber", "JsonString" and "JsonBoolean" classes since the distinction becomes irrelevant in some cases such as when a CLR number can be deserialized from a JSON number or from a JSON string (for quoted numbers) and similarly be serialized as a JSON string or a JSON number depending on options. However, it is still possible to obtain the JSON token type for a node after deserialization.
- Supports custom and built-in serialization converters. Has a reference to the options class to make this possible. Thus it can leverage the many built-in converters (for value types, collections and objects) and any registered custom converters to do this.
  - Due to this, there is currently no explicit\implicit casts or known-type-specific conversions (e.g. no one-off method to convert to\from a `System.Double`). 
  - Instead of one-off methods for known types, a `JsonNode.Convert<T>()` method is used which may perform non-trivial conversions such as converting a `JsonValue` with an internal `byte[]` (obtained from raw UTF8 JSON) to a POCO, or converting a string (obtained from a JSON string) to a `System.Double`.

It is actively been reviewed, so the above is subject to change.

Once the API is internally approved, we should do new usability study on it and iterate the API if necessary.